### PR TITLE
Adding custom report filter

### PR DESCRIPTION
**Pull request by Jasmin-1993**
Sometimes we need a custom filter to filter the reported information to reduce the storage pressure

### CODE REVIEW

**Comment by wu-sheng** [comment]
Storage pressure should be taken care of by backend OAP capability.
If you don't want to trace at the beginning, a sampler makes a more sense before creating context.

**Comment by Jasmin-1993** [comment]
for example: only submit requests below 100ms

**Comment by wu-sheng** [comment]
That is actually very tricky. You could make a lot of VNode when visuals trace.
Sampling usually is a tradeoff, but your point of reducing storage doesn't make sense to provide this feature.

**Comment by Jasmin-1993** [comment]
some of the problems just mentioned should actually reduce the pressure on grpc to report

**Comment by Jasmin-1993** [comment]
In the case of high QPS, grpc server needs a lot of resources. In fact, we can only collect and analyze requests with more than 100ms or errors

**Comment by wu-sheng** [comment]
But your way as post sampling actually could make things not as expected in the runtime.
When your service has performance impact, the sampler would report more and more segments, which slows the service more.
This is not a good practice in prod.

**Comment by wu-sheng** [comment]
> In the case of high QPS, grpc server needs a lot of resources. In fact, we can only collect and analyze requests with more than 100ms or errors

How many QPS are you talking about? Do you really test the tracing cost VS reporting cost?

In most languages, tracing cost is almost 5-10 times than pure reporting.

**Comment by Jasmin-1993** [comment]
In fact, the problem we encounter is that in 50000qps requests, we only need to track those wrong and slow requests, and the vast majority of requests (99.9%) do not need to be submitted to the tracking. If the full number of grpc submissions, the server received by grpc may need more resources, so we want to filter out (retain wrong and slow requests) when submitting to reduce the pressure of grpc reporting. Do you have a better approach?

**Comment by wu-sheng** [comment]
But that is not SkyWalking's typical use case. I am actually confused about your decision.
If you just need some traces of error or slow, and don't care about the analysis/metrics, why do you choose SkyWalking rather than many others out there?
SkyWalking's traces are designed for dependencies and metrics analysis, but it seems this is also not your purpose.
Then why do you choose SkyWalking, but use it like a pure tracing stack?

**Comment by wu-sheng** [comment]
Back to the codes, I think you need to update the doc at https://github.com/SkyAPM/go2sky#configuration about how to use this. 

We need the doc to explain what is `CustomReportStrategy`. But first, this should be renamed to `ReportStrategy`.(`Custom` is not necessary).

File: README.md [hunk 1]
```diff
@@ -32,6 +32,22 @@ if err != nil {
 }
 defer r.Close()
 tracer, err := go2sky.NewTracer("example", go2sky.WithReporter(r))
+```
+In some scenarios, we may need a filter to filter segments that do not need to be submitted to reduce the pressure of grpc reporting. For example, we just need to track the request of error.
```
Comment (wu-sheng) [inline README.md:36]:
```suggestion
In some scenarios, we may need a filter to filter segments that do not need to be submitted, for example, to reduce the load of gRPC reporting, or only track the request of error.
```

**Comment by wu-sheng** [comment]
@arugal Could you review the codes? Although I am not supporting this use case, this seems harmless either.

I will leave the judgment for you.

**Comment by Jasmin-1993** [comment]
Has been updated

File: reporter/grpc.go [hunk 2]
```diff
@@ -232,6 +236,9 @@ func (r *gRPCReporter) initSendPipeline() {
 				continue StreamLoop
 			}
 			for s := range r.sendCh {
+				if !r.rs(s) {
+					continue
+				}
```
Comment (arugal) [inline reporter/grpc.go:241]:
```suggestion
				if r.rs != nil && !r.rs(s) {
					continue
				} 
```

I recommend changing it to this, for two reasons:
1. No need to set defaul `ReportStrategy`.
2. Avoid panic caused by user setting `nil`.

Comment (Jasmin-1993) [reply reporter/grpc.go]: Has been updated

**Review by arugal — APPROVED** [review]
LGTM

**Pull request merged**